### PR TITLE
feat(spanner): add samples for using multi-region encryption keys for backups and database

### PR DESCRIPTION
**Pull request by rahul2393**
## Description

Reference PRs:
* https://github.com/GoogleCloudPlatform/dotnet-docs-samples/pull/2832
* https://github.com/GoogleCloudPlatform/php-docs-samples/pull/2044
* https://github.com/googleapis/python-spanner/pull/1191
* https://github.com/googleapis/nodejs-spanner/pull/2099
* https://github.com/googleapis/java-spanner/pull/3524

Note: Before submitting a pull request, please open an issue for discussion if you are not associated with Google.

## Checklist
- [x] I have followed [Contributing Guidelines from CONTRIBUTING.MD](https://github.com/GoogleCloudPlatform/golang-samples/blob/main/CONTRIBUTING.md)
- [ ] **Tests** pass:   `go test -v ./..` (see [Testing](https://github.com/GoogleCloudPlatform/golang-samples/blob/main/CONTRIBUTING.md#testing))
- [x] **Code formatted**:   `gofmt` (see [Formatting](https://github.com/GoogleCloudPlatform/golang-samples/blob/main/CONTRIBUTING.md#formatting))
- [x] **Vetting** pass:   `go vet` (see [Formatting](https://github.com/GoogleCloudPlatform/golang-samples/blob/main/CONTRIBUTING.md#formatting))
- [ ] These samples need a new **API enabled** in testing projects to pass (let us know which ones)
- [ ] These samples need a new/updated **env vars** in testing projects set to pass (let us know which ones)
- [ ] This sample adds a new sample directory, and I updated the [CODEOWNERS file](https://github.com/GoogleCloudPlatform/golang-samples/blob/main/.github/CODEOWNERS) with the codeowners for this sample
- [ ] This sample adds a new **Product API**, and I updated the [Blunderbuss issue/PR auto-assigner](https://github.com/GoogleCloudPlatform/golang-samples/blob/main/.github/blunderbuss.yml) with the codeowners for this sample
- [x] Please **merge** this PR for me once it is approved

### CODE REVIEW

**Review by harshachinta — APPROVED** [review]

File: spanner/spanner_snippets/spanner/spanner_create_backup_with_multi_region_encryption_key.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+	"fmt"
+	"io"
+	"regexp"
+	"time"
+
+	database "cloud.google.com/go/spanner/admin/database/apiv1"
+	adminpb "cloud.google.com/go/spanner/admin/database/apiv1/databasepb"
+	pbt "github.com/golang/protobuf/ptypes/timestamp"
+)
+
```
Comment (telpirion) [inline spanner/spanner_snippets/spanner/spanner_create_backup_with_multi_region_encryption_key.go:30]: nit: add function-level Godoc

File: spanner/spanner_snippets/spanner/spanner_create_backup_with_multi_region_encryption_key.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+func createBackupWithCustomerManagedMultiRegionEncryptionKey(ctx context.Context, w io.Writer, db, backupID string, kmsKeyNames []string) error {
+	// db = `projects/<project>/instances/<instance-id>/database/<database-id>`
+	// backupID = `my-backup-id`
+	// kmsKeyNames := []string{"projects/my-project/locations/locations/<location1>/keyRings/<keyRing>/cryptoKeys/<keyId>",
+	//	 "projects/my-project/locations/locations/<location2>/keyRings/<keyRing>/cryptoKeys/<keyId>",
+	//	 "projects/my-project/locations/locations/<location3>/keyRings/<keyRing>/cryptoKeys/<keyId>",
+	// }
+	matches := regexp.MustCompile("^(.+)/databases/(.+)$").FindStringSubmatch(db)
```
Comment (telpirion) [inline spanner/spanner_snippets/spanner/spanner_create_backup_with_multi_region_encryption_key.go:38]: issue: pass in the instance name as an argument to this function. The regex isn't required for demonstrating this method call.

Here and below.

File: spanner/spanner_snippets/spanner/spanner_create_database_with_multi_region_encryption_key.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+import (
+	"context"
+	"fmt"
+	"io"
+	"regexp"
+
+	database "cloud.google.com/go/spanner/admin/database/apiv1"
+	adminpb "cloud.google.com/go/spanner/admin/database/apiv1/databasepb"
+)
+
```
Comment (telpirion) [inline spanner/spanner_snippets/spanner/spanner_create_database_with_multi_region_encryption_key.go:28]: nit: add function-level Godoc.

File: spanner/spanner_snippets/spanner/spanner_create_database_with_multi_region_encryption_key.go [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+func createDatabaseWithCustomerManagedMultiRegionEncryptionKey(ctx context.Context, w io.Writer, db string, kmsKeyNames []string) error {
+	// db = `projects/<project>/instances/<instance-id>/database/<database-id>`
+	// kmsKeyNames := []string{"projects/my-project/locations/locations/<location1>/keyRings/<keyRing>/cryptoKeys/<keyId>",
+	//	 "projects/my-project/locations/locations/<location2>/keyRings/<keyRing>/cryptoKeys/<keyId>",
+	//	 "projects/my-project/locations/locations/<location3>/keyRings/<keyRing>/cryptoKeys/<keyId>",
+	// }
+	matches := regexp.MustCompile("^(.+)/databases/(.+)$").FindStringSubmatch(db)
```
Comment (telpirion) [inline spanner/spanner_snippets/spanner/spanner_create_database_with_multi_region_encryption_key.go:35]: issue: similar to previous--pass in instance and DB name as individual arguments. Remove the regex matching.

File: spanner/spanner_snippets/spanner/spanner_restore_backup_with_multi_region_encryption_key.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+import (
+	"context"
+	"fmt"
+	"io"
+	"regexp"
+
+	database "cloud.google.com/go/spanner/admin/database/apiv1"
+	adminpb "cloud.google.com/go/spanner/admin/database/apiv1/databasepb"
+)
+
```
Comment (telpirion) [inline spanner/spanner_snippets/spanner/spanner_restore_backup_with_multi_region_encryption_key.go:28]: nit: add function-level Godoc.

File: spanner/spanner_snippets/spanner/spanner_restore_backup_with_multi_region_encryption_key.go [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+func restoreBackupWithCustomerManagedMultiRegionEncryptionKey(ctx context.Context, w io.Writer, db, backupID string, kmsKeyNames []string) error {
+	// db = `projects/<project>/instances/<instance-id>/database/<database-id>`
+	// backupID = `my-backup-id`
+	// kmsKeyName = `projects/<project>/locations/<location>/keyRings/<key_ring>/cryptoKeys/<kms_key_name>`
+	matches := regexp.MustCompile("^(.*)/databases/(.*)$").FindStringSubmatch(db)
```
Comment (telpirion) [inline spanner/spanner_snippets/spanner/spanner_restore_backup_with_multi_region_encryption_key.go:33]: issue: remove regex matching code and pass arguments into func.

File: spanner/spanner_snippets/spanner/spanner_restore_backup_with_multi_region_encryption_key.go [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+	})
+	if err != nil {
+		return fmt.Errorf("restoreBackupWithCustomerManagedMultiRegionEncryptionKey.RestoreDatabase: %w", err)
+	}
+	// Wait for restore operation to complete.
+	restoredDatabase, err := restoreOp.Wait(ctx)
+	if err != nil {
+		return fmt.Errorf("restoreBackupWithCustomerManagedMultiRegionEncryptionKey.Wait: %w", err)
+	}
+	// Get the information from the newly restored database.
```
Comment (telpirion) [inline spanner/spanner_snippets/spanner/spanner_restore_backup_with_multi_region_encryption_key.go:67]: nit: add a space before this comment. As written, this sample presents a "wall of text."

**Review by telpirion — CHANGES_REQUESTED** [review]
Overall this looks good! Some comments left. Consider adding a blank line inbetween logical groups of code.

File: spanner/spanner_snippets/spanner/spanner_create_backup_with_multi_region_encryption_key.go (hunk 1 above)
Comment (rahul2393) [reply spanner/spanner_snippets/spanner/spanner_create_backup_with_multi_region_encryption_key.go:29]: Done

File: spanner/spanner_snippets/spanner/spanner_restore_backup_with_multi_region_encryption_key.go (hunk 5 above)
Comment (rahul2393) [reply spanner/spanner_snippets/spanner/spanner_restore_backup_with_multi_region_encryption_key.go:27]: Done

File: spanner/spanner_snippets/spanner/spanner_restore_backup_with_multi_region_encryption_key.go (hunk 7 above)
Comment (rahul2393) [reply spanner/spanner_snippets/spanner/spanner_restore_backup_with_multi_region_encryption_key.go:66]: Done

**Pull request merged**